airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeDestinationHandler.kt

```diff
@@ -0,0 +1,450 @@
+        return existingTable.columns.containsKey(abRawIdColumnName) &&
+            toJdbcTypeName(AirbyteProtocolType.STRING) ==
+                existingTable.columns[abRawIdColumnName]!!.type
+    }
+
+    override fun isAirbyteExtractedAtColumnMatch(existingTable: TableDefinition): Boolean {
+        val abExtractedAtColumnName: String =
+            JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT.uppercase(Locale.getDefault())
+        return existingTable.columns.containsKey(abExtractedAtColumnName) &&
+            toJdbcTypeName(AirbyteProtocolType.TIMESTAMP_WITH_TIMEZONE) ==
+                existingTable.columns[abExtractedAtColumnName]!!.type
+    }
+
+    override fun isAirbyteMetaColumnMatch(existingTable: TableDefinition): Boolean {
+        val abMetaColumnName: String =
+            JavaBaseConstants.COLUMN_NAME_AB_META.uppercase(Locale.getDefault())
+        return existingTable.columns.containsKey(abMetaColumnName) &&
+            "VARIANT" == existingTable.columns[abMetaColumnName]!!.type
+    }
+
+    override fun existingSchemaMatchesStreamConfig(
+        stream: StreamConfig?,
+        existingTable: TableDefinition
+    ): Boolean {
+        val pks = getPks(stream)
+        // This is same as JdbcDestinationHandler#existingSchemaMatchesStreamConfig with upper case
+        // conversion.
+        // TODO: Unify this using name transformer or something.
+        if (
+            !isAirbyteRawIdColumnMatch(existingTable) ||
+                !isAirbyteExtractedAtColumnMatch(existingTable) ||
+                !isAirbyteMetaColumnMatch(existingTable)
+        ) {
+            // Missing AB meta columns from final table, we need them to do proper T+D so trigger
+            // soft-reset
+            return false
+        }
+        // Rewritten in Kotlin
+        // https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeDestinationHandler.java#L270-L287
+        val intendedColumns: LinkedHashMap<String, String> =
+            stream!!.columns.asSequence().associateTo(LinkedHashMap()) {
+                it.key.name to toJdbcTypeName(it.value)
+            }
+        // Filter out Meta columns since they don't exist in stream config.
+        val actualColumns: LinkedHashMap<String, String> =
+            existingTable.columns
+                .filter { it ->
+                    V2_FINAL_TABLE_METADATA_COLUMNS.map { it.uppercase() }
+                        .none { abCol -> abCol == it.key }
+                }
+                .asSequence()
+                .associateTo(LinkedHashMap()) { it.key!! to it.value.type }
+        // soft-resetting https://github.com/airbytehq/airbyte/pull/31082
+        val hasPksWithNonNullConstraint: Boolean =
+            existingTable.columns.any { pks.contains(it.key) && !it.value.isNullable }
```

Comment:
Line 301 - 316 is rewrite of this
https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeDestinationHandler.java#L270-L287